### PR TITLE
Multiple model endpoints

### LINKED ISSUE #175: Multiple model invocation endpoints

### PR DESCRIPTION
**Pull request by kirillmakhonin**
This closes #175, #174, #50

### CODE REVIEW

**Comment by kirillmakhonin** [comment]
test this please

**Comment by alexander-semenets** [comment]
LGTM

**Comment by kirillmakhonin** [comment]
@alexander-semenets you haven't set review status

**Comment by kirillmakhonin** [comment]
@aliaksandr-d @akharlamov awaiting your reviews

File: legion/legion/edi/deploy.py [hunk 1]
```diff
@@ -56,10 +56,10 @@ def build_model(args):
...
-        with legion.io.ModelContainer(external_reader.path, do_not_load_model=True) as container:
-            model_id = container.get('model.id', None)
-            if args.model_id:
-                model_id = args.model_id
+        container = legion.io.ModelContainer().load(external_reader.path)
+        model_id = container.model_id
+        if args.model_id:
+            model_id = args.model_id
 
         if not model_id:
```
Comment (kirillmakhonin) [inline legion/legion/edi/deploy.py]: Unreachable code

File: legion/legion/serving/pyserve.py [hunk 2]
```diff
@@ -135,14 +204,15 @@ def init_model(application):
...
     :return: model instance
     """
-    if 'MODEL_FILE' in application.config:
-        file = application.config['MODEL_FILE']
-        LOGGER.info("Loading model from %s", file)
-        with legion.io.ModelContainer(file) as container:
-            model = container.model
-    else:
+    if 'MODEL_FILE' not in application.config:
         raise Exception('Unknown model file')
```
Comment (akharlamov) [inline legion/legion/serving/pyserve.py]: 'No model file provided'

File: legion/legion/utils.py [hunk 3]
```diff
@@ -450,7 +453,7 @@ def send_header_to_stderr(header, value):
     :type value: str
     :return: None
     """
-    message = 'X-Legion-%s:%s' % (header, value)
```
Comment (akharlamov) [inline legion/legion/utils.py]: I think we should put X-Legion- prefix to name of headers themselves. So, nobody need to look it up through the code.

File: legion/legion/io.py [hunk 4]
```diff
@@ -65,276 +68,448 @@ def _get_column_types(param_types):
     return deduct_types_on_pandas_df(data_frame=pandas_df_sample, extra_columns=custom_props)
 
 
-class ModelContainer:
+class ModelEndpoint:
```
Comment (akharlamov) [inline legion/legion/io.py]: PyModel ?

File: legion/legion/edi/deploy.py (hunk 1 above)
Comment (kirillmakhonin) [reply legion/legion/edi/deploy.py]: This has been already fixed in #225

File: legion/legion/edi/deploy.py [hunk 5]
```diff
@@ -56,10 +56,10 @@ def build_model(args):
...
             raise Exception('Model ID is missed')
```
[the remark this answers is not in the document]
Comment (kirillmakhonin) [reply legion/legion/edi/deploy.py]: This has been already fixed in #225

File: legion/legion/utils.py (hunk 3 above)
Comment (kirillmakhonin) [reply legion/legion/utils.py]: It has been fixed in #225

File: examples/MovieLens/surprise.v3.py [hunk 6]
```diff
@@ -172,13 +172,11 @@ def parse_input(input):
     def recommend(input):
         return top3_recommendations[input['uid']]
 
-legion.io.export(
+legion.io.ModelContainer().export(
     recommend,
     {
         'uid': legion.model.types.ColumnInformation(legion.model.types.Integer, legion.model.int32)
```
[the remark this answers is not in the document]
Comment (kirillmakhonin) [reply examples/MovieLens/surprise.v3.py]: Will do

File: legion/docs/source/model_rest_api.md [hunk 7]
```diff
@@ -1,37 +1,268 @@
...
+                        "type": "Integer"
+                    }
+                },
+                "name": "default",
+                "use_df": true
+            }
+        },
+        "id": "income",
+        "version": "1.1"
+    }
```
[the remark this answers is not in the document]
Comment (kirillmakhonin) [reply legion/docs/source/model_rest_api.md]: Will do

File: legion/legion/serving/pyserve.py (hunk 2 above)
Comment (kirillmakhonin) [reply legion/legion/serving/pyserve.py]: Will do

**Pull request merged**